components/encryption/src/master_key/kms.rs

```diff
@@ -0,0 +1,306 @@
+use rusoto_core::request::DispatchSignedRequest;
+use rusoto_core::request::HttpClient;
+use rusoto_credential::{DefaultCredentialsProvider, StaticProvider};
+use rusoto_kms::{DecryptRequest, GenerateDataKeyRequest, Kms, KmsClient};
+use tokio::runtime::{Builder, Runtime};
+
+use super::{metadata::MetadataKey, Backend, FileBackend, PlainTextBackend, WithMetadata};
+use crate::config::KmsConfig;
+use crate::encrypted_file::EncryptedFile;
+use crate::{Error, Result};
+
+const KMS_ENCRYPTION_KEY_NAME: &str = "kms_encryption.key";
+
+// Always use AES 256 for encrypting master key.
+const KMS_DATA_KEY_METHOD: EncryptionMethod = EncryptionMethod::Aes256Ctr;
+const AWS_KMS_DATA_KEY_SPEC: &str = "AES_256";
+const AWS_KMS_VENDOR_NAME: &[u8] = b"AWS";
+
+struct AwsKms {
+    client: KmsClient,
+    current_key_id: String,
+}
+
+impl AwsKms {
+    fn new(config: KmsConfig) -> Result<AwsKms> {
+        let http_dispatcher = HttpClient::new().unwrap();
+
+        AwsKms::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    // TODO following code is almost the same as external_storage s3 client,
+    //      should be wrapped together.
+    fn with_request_dispatcher<D>(config: KmsConfig, dispatcher: D) -> Result<AwsKms>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        if config.key_id.is_empty() {
+            return Err(Error::Other(
+                "KMS key id can not be empty".to_owned().into(),
+            ));
+        }
+
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::Other(format!("invalid region format {}: {}", config.region, e).into())
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region,
+                endpoint: config.endpoint,
+            }
+        };
+
+        let client = if config.access_key.is_empty() || config.secret_access_key.is_empty() {
+            let cred_provider = DefaultCredentialsProvider::new()
+                .map_err(|e| Error::Other(format!("unable to get credentials: {}", e).into()))?;
+            KmsClient::new_with(dispatcher, cred_provider, region)
+        } else {
+            let cred_provider = StaticProvider::new(
+                config.access_key,
+                config.secret_access_key,
+                None, /* token */
+                None, /* valid_for */
+            );
+            KmsClient::new_with(dispatcher, cred_provider, region)
+        };
+
+        Ok(AwsKms {
+            client,
+            current_key_id: config.key_id,
+        })
+    }
+
+    fn decrypt(&self, runtime: &mut Runtime, ciphertext: &[u8]) -> Result<Vec<u8>> {
+        let decrypt_request = DecryptRequest {
+            ciphertext_blob: ciphertext.to_vec().into(),
+            encryption_context: None,
+            grant_tokens: None,
+            encryption_algorithm: None,
+            key_id: None,
+        };
```

Comment:
nit: simply fill with default? `DecryptRequest { ciphertext_blob: ..., ..Default::default() }`. Same for `GenerateDataKeyRequest`.

Reply:
We could, but I think it's better to keep it explicit.

Reply:
make sense, and let's add some inline comments whenever it is worth noting. e.g. `encryption_algorithm` is none means it is using `SYMMETRIC_DEFAULT`, `key_id` is none means we are using `key_id` encoded in ciphertext.